### PR TITLE
[Python] distinguish decorator at beginning of line from __matmul__

### LINKED ISSUE #3957: [Python mode] wrong highlight for  `@` as binary operator
In the following, `@` should be recognized as binary operator, and not decorator.
```python
x = a@b@c
```

Here the decorator highlighting is correct. 

```python
@b
def function():
    pass
```

Screnshot:
screen shot 2016-04-13 at 10 09 28

Likely doable  by having the decorator recognized only as beginning of line.

I'll try to have a look, but opening an issue first. 

Thanks !

### PR DESCRIPTION
**Pull request by Carreau**
`@` as __matmul__ operator only on Python 3,

Add tests for Python mode for this use case.

Closes #3957

### CODE REVIEW

File: mode/python/python.js [hunk 1]
```diff
@@ -249,15 +249,24 @@
     }
 
     function tokenLexer(stream, state) {
+      var sol = stream.sol();
+      if (sol){
+          state.whiteSpaceAtBeginningOfLine = true;
+      }
```
Comment (marijnh) [inline mode/python/python.js]: The project uses two spaces for indentation. And you can directly put the call to `sol()` in the condition, no need for a variable.

File: mode/python/python.js [hunk 2]
```diff
@@ -249,15 +249,24 @@
...
       var style = state.tokenize(stream, state);
       var current = stream.current();
 
       // Handle decorators
-      if (current == "@") {
-        if (parserConf.version && parseInt(parserConf.version, 10) == 3)
-          return stream.match(identifiers, false) ? "meta" : "operator";
-        else
-          return stream.match(identifiers, false) ? "meta" : ERRORCLASS;
+      if (state.whiteSpaceAtBeginningOfLine){
```
Comment (marijnh) [inline mode/python/python.js]: You can just add this to the condition of the existing `if` with `&&`. No need to add another wrapping `if`

File: mode/python/python.js [hunk 3]
```diff
@@ -249,15 +249,24 @@
...
+          if (current == "@") {
+            if (parserConf.version && parseInt(parserConf.version, 10) == 3)
+              return stream.match(identifiers, false) ? "meta" : "operator";
+            else
+              return stream.match(identifiers, false) ? "meta" : ERRORCLASS;
+          }
+      }
+      if(!current.match(/^ +$/, false)){
+          state.whiteSpaceAtBeginningOfLine = false;
```
Comment (marijnh) [inline mode/python/python.js]: (two spaces)

File: mode/python/python.js (hunk 1 above)
Comment (Carreau) [inline mode/python/python.js]: > The project uses two spaces for indentation

Ooops, sorry, I knew that. 



> And you can directly put the call to sol() in the condition, no need for a variable.

Good point.

File: mode/python/python.js (hunk 2 above)
Comment (Carreau) [inline mode/python/python.js]: done.

File: mode/python/python.js (hunk 3 above)
Comment (Carreau) [inline mode/python/python.js]: Done.

File: mode/python/test.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+     "    [meta @dec]",
+     "    [keyword def] [def function]():",
+     "        [keyword pass]");
+
+  MT("matmulWithSpace:", "[variable a] [operator @] [variable b]");
+  MT("matmulWithoutSpace:", "[variable a][operator @][variable b]");
+  MT("matmulSpaceBefore:", "[variable a] [operator @][variable b]");
+
+  MT("fValidStringPrefix", "[string f'this is a {formatted} string']");
+  MT("uValidStringPrefix", "[string u'this is an unicode string']");
```
Comment (Carreau) [inline mode/python/test.js]: Also added 2 test for  #3975 as 03d684f you merged recently, and rebased on master.

**Comment by marijnh** [comment]
I've made the cleanups myself in 3b36e58 so that this can go into today's release. Merged as db62891c78

**Comment by marijnh** [comment]
Oh, oops, our messages crossed each other there (and somehow github didn't do its usual auto-updating on the thread). Sorry for the duplicate work!

**Comment by marijnh** [comment]
I'll see if I can pull out the new tests and merge them separately.

**Comment by marijnh** [comment]
Done in a579a10

**Comment by Carreau** [comment]
> Oh, oops, our messages crossed each other there (and somehow github didn't do its usual auto-updating on the thread). Sorry for the duplicate work!

> I'll see if I can pull out the new tests and merge them separately.

No worries I could have submitted another patch. Thanks !

**Pull request closed without merging**